tracer/src/Datadog.Tracer.Native/fault_tolerant_rewriter.cpp

```diff
@@ -0,0 +1,594 @@
+    const auto instrumentationVersion = m_methodRewriter->GetInstrumentationVersion(moduleHandler, methodHandler);
+    const auto instrumentingProduct = m_methodRewriter->GetInstrumentingProduct(moduleHandler, methodHandler);
+    FaultTolerantTokens* faultTolerantTokens = moduleHandler->GetModuleMetadata()->GetFaultTolerantTokens();
+    faultTolerantTokens->EnsureCorLibTokens();
+
+    auto instrumentedMethodName = caller->name + WStr("<Instrumented>");
+    instrumentedMethodName.erase(std::remove(instrumentedMethodName.begin(), instrumentedMethodName.end(), L'.'),
+                                 instrumentedMethodName.end());
+    instrumentedMethodName.erase(std::remove(instrumentedMethodName.begin(), instrumentedMethodName.end(), L'_'),
+                                 instrumentedMethodName.end());
+    auto originalMethodName = caller->name + WStr("<Original>");
+    originalMethodName.erase(std::remove(originalMethodName.begin(), originalMethodName.end(), L'.'),
+                             originalMethodName.end());
+    originalMethodName.erase(std::remove(originalMethodName.begin(), originalMethodName.end(), L'_'),
+                             originalMethodName.end());
+
+    // Is the type generic:
+    auto metadataImport = moduleHandler->GetModuleMetadata()->metadata_import;
+    auto isGenericOrNestedType = false;
+    auto argGenericCount = 0;
+
+    if (caller->type.isGeneric)
+    {
+        isGenericOrNestedType = true;
+        int number = std::stoi(caller->type.name.substr(caller->type.name.find(L'`') + 1));
+        argGenericCount += number;
+    }
+
+    auto currentType = caller->type.parent_type;
+    while (currentType != nullptr)
+    {
+        if (currentType->isGeneric)
+        {
+            isGenericOrNestedType = true;
+            int number = std::stoi(currentType->name.substr(currentType->name.find(L'`') + 1));
+            argGenericCount += number;
+        }
+
+        currentType = currentType->parent_type;
+    }
+
+    if (isGenericOrNestedType)
+    {
+        unsigned typeBuffer;
+        auto typeSize = CorSigCompressToken(caller->type.id, &typeBuffer);
+        COR_SIGNATURE signature[500];
+        unsigned offset = 0;
+        unsigned argCountBuffer;
+        auto argCountSize = CorSigCompressData(argGenericCount, &argCountBuffer);
+        mdTypeSpec containingTypeSpec = mdTypeSpecNil;
+
+        signature[offset++] = ELEMENT_TYPE_GENERICINST;
+        signature[offset++] = caller->type.valueType ? ELEMENT_TYPE_VALUETYPE : ELEMENT_TYPE_CLASS;
+        memcpy(&signature[offset], &typeBuffer, typeSize);
+        offset += typeSize;
+        memcpy(&signature[offset], &argCountBuffer, argCountSize);
+        offset += argCountSize;
+
+        for (int genArgIndex = 0; genArgIndex < argGenericCount; genArgIndex++)
+        {
+            signature[offset++] = ELEMENT_TYPE_VAR;
+            signature[offset++] = genArgIndex;
```

Comment:
Are there any combinations of generics/parent types where this doesn't work? Or does this work for all cases?

Reply:
From the exploration testings I've seen crazy combinations of nested generic types, enclosed type in a generic, multiple nestings of generic types and it worked in all cases.
Just as a note, in the code block you've selected I'm creating the type spec token that represents the call to be made. If you deal with open generic types then the signature need to have: `ELEMENT_TYPE_VAR` x `CountOfGenericTypesIncludingNestings`.